mkdocs/docs/bindings/c-api.md

```diff
@@ -1,9 +1,224 @@
 
 ## Prerequisites
 
-<!-- TODO(??): build from source, vcpkg -->
+To use IREE's C API, you will need to build the runtime
+[from source](../building-from-source/getting-started.md). The
+[iree-template-cpp](https://github.com/iml130/iree-template-cpp) community
+project also shows how to integrate IREE into an external project using
+CMake[^2].
+
+## Concepts
+
+By default, IREE uses a Virtual Machine (VM) at runtime to interpret program
+instructions on the host system. VM instructions may also be lowered further to
+LLVM IR, C, or other representations for static or resource constrained
+deployment.
+
+The VM supports generic operations like loads, stores, arithmetic, function
+calls, and control flow. It builds streams of more complex program logic and
+dense math into command buffers that are dispatched to hardware backends
+through the Hardware Abstraction Layer (HAL) interface.
+
+Most interaction with IREE's C API involves either the VM or the HAL.
+
+<!-- TODO(scotttodd): diagrams -->
+
+### IREE VM
+
+* VM _instances_ can serve multiple isolated execution _contexts_
+* VM _contexts_ are effectively sandboxes for loading modules and running
+  programs
+* VM _modules_ provide extra functionality to execution _contexts_, such as
+  access to hardware accelerators through the HAL
+
+### IREE HAL
+
+* HAL _drivers_ are used to enumerate and create HAL _devices_
+* HAL _devices_ interface with hardware, such as by allocating device memory,
+  preparing executables, recording and dispatching command buffers, and
+  synchronizing with the host
+
+## Using the C API
+
+### Setup
+
+Include headers:
+
+``` c
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/vm/api.h"
+
+// The VM bytecode and HAL modules will typically be included, along
+// with those for the specific HAL drivers your application uses.
+// Functionality extensions can be used via custom modules.
+#include "iree/modules/hal/hal_module.h"
+#include "iree/hal/dylib/registration/driver_module.h"
+#include "iree/vm/bytecode_module.h"
+```
+
+Check the API version and register components:
+
+``` c
+// Statically linking the runtime should never have version conflicts,
+// however dynamically linking against a runtime shared object must
+// always verify that the version is as expected.
+iree_api_version_t actual_version;
+IREE_CHECK_OK(iree_api_version_check(IREE_API_VERSION_LATEST, &actual_version));
+
+// Modules with custom types must be statically registered before use.
+IREE_CHECK_OK(iree_hal_module_register_types());
+
+// Device drivers are managed through registries.
+// Applications may use multiple registries to more finely control driver
+// lifetimes and visibility.
+IREE_CHECK_OK(iree_hal_dylib_driver_module_register(
+    iree_hal_driver_registry_default()));
+```
+
+!!! tip
+    Macros like `IREE_CHECK_OK()` can be used to help with error handling.
+    In many cases, errors should be bubbled up to hosting applications to
```

Comment:
worth mentioning that this is like `exit(1)` and being a bit stronger: "Applications should propagate errors and handle or report them as desired."